// @gdo-tvstack/dan-auth-middleware/auth/model/Role.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class IRole {
}
exports.IRole = IRole;
class Role {
    constructor(role) {
        this._id = role._id;
        this.level = role.level;
        this.type = role.type;
        this.code = role.code;
        this.parent = role.parent;
    }
    loopProtectedDescendsFrom(origin, ancestor, rolesDomain) {
        const parents = rolesDomain.filter((candidate) => candidate._id === this.parent);
        const parent = parents && parents.length > 0 ? parents[0] : null;
        if (parent && parent._id === origin._id) {
            const message = "Found a loop in the role hierarchy";
            console.log(message);
            throw { message };
        }
        return (this._id === ancestor._id) || (parent && parent.loopProtectedDescendsFrom(origin, ancestor, rolesDomain));
    }
    descendsFrom(ancestor, rolesDomain) {
        return this.loopProtectedDescendsFrom(this, ancestor, rolesDomain);
    }
}
exports.default = Role;
